### ISSUE TITLE
Addressing object keys

### ISSUE THREAD

**Issue by benkehoe** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#156]
Please excuse me if this has been fully covered; I've been reading over the history of #109, [the minutes where it was discussed](https://datatracker.ietf.org/doc/minutes-112-jsonpath/), and messages in the [mailing list](https://mailarchive.ietf.org/arch/msg/jsonpath/MT_WROl1XpUWYOGy69HMAR2DVVI/) and I'm not sure this specific question has been asked.

There has been discussion of how object keys might be referenced in a filter selector (e.g., with an `index()` function), which it seems like has been settled as out of scope. A different question, to my mind, is "Can a JSON Path address an object key?"—has this been asked? Is an object key a node? The spec says:

> For the purposes of this specification, a value as defined by RFC8259 is also viewed as a tree of nodes.  Each node, in turn, holds a value.  Further nodes within each value are the elements of arrays and the member values of objects and are themselves values. (The type of the value held by a node may also be referred to as the type of the node.)

This does seem to scope nodes only to values. The closest thing I could find to someone asking my question is [this message](https://mailarchive.ietf.org/arch/msg/jsonpath/iIej1cnzEoHkjKgBO41ZXOXhUWI/) which mentions key-value pairs, and has no further discussion. It seems a little strange to me that as general-purpose as JSON Path is, it does not provide for addressing all elements of a JSON document.

One possible answer to this is, "an implementation may make the object key available as a property of a node in the nodelist it returns for a path". This would suffice for a user processing a JSON document and needing to extract this information, but that's separate from being able to reference that key as a first-class element within the document, for example when an interface accepts a JSON pointer as input for some action it will take on the pointed-to element.

Another possible answer, to my mind, would be something like an "index name selector". For the sake of argument, say this took the form of `#index` (setting aside choice of separator, etc), and the path definition was something like 
```abnf
json-path = root-selector *(S primary-selector) [ S primary-selector S "#index" ] ; the root can't have an index name
                                                                                  ; no further selectors can be applied after it
primary-selector = (dot-selector        /
                    dot-wild-selector   /
                    index-selector      /
                    index-wild-selector /
                    list-selector       /
                    slice-selector      /
                    descendant-selector /
                    filter-selector)
```
The index name selector would be defined to produce a string node whose value is the object key or array index when applied to a node in an object or array, or no node otherwise. With something like this, it would allow a user to definitively reference object keys (and array indexes) with a JSON Path. And it would additionally allow object keys to be referenced in filter selectors without a separate syntax, which would address a number of the concerns I saw surrounding that.

As I said, this may have been thought about and decided upon already, but as I couldn't find it after some digging, I thought I would pose the question for consideration.

**Comment by nichtich** [comment]
The `index()` function is included in examples but seems not to be standardized yet. A simple `index` instead of `index(@)` would work as well and be cleaner:

    $[?index == "key"]
    $[?index > 1]

**Comment by benkehoe** [comment]
That is slightly different. For the document `{"key": "value"}`, I would expect `$[?index == "key"]` to resolve to `"value"`. The question I have is, could there be a JSON Path that would resolve, for that document, to `"key"`? The strawman proposal above would say that appending `#index` onto a path would return the index of the node(s) that are returned by the path. It's not very interesting for that example, where `$[?index == "key"]#index` would return `"key"`; a more interesting case would be `$[?(@ == "value")]#index` would return `"key"`

**Comment by timbray** [comment]
This is a real problem with JSONPath and for the moment we don't have a solution.  Closing for now.

**Comment by gregsdennis** [comment]
Making a note that this feature (or one like it) was recently asked about in a [StackOverflow question](https://stackoverflow.com/q/75187796/878701).

**Comment by benkehoe** [comment]
I'd be glad for you all to take another look, but I think it's orthogonal to functions since they can only be used inside filter selectors. You could have a function `index()` that has an argument of `OptionalNode` and a return value of `OptionalValue`, where the value is the key/index of the node (if it has one), but that return value, which exists in the document, remains unaddressable as a JSONPath.

I think it's a different use case from that StackOverflow question. They have:
```json
{ 
    "a" : 1,
    "key_to_find" : "value_of_key",
    "nested" : {
         "value_of_key" : "real_value",
         "b" : 2
    }
}
```
and want to reference the value `"real_value"`. So in theory they could do something like `$.nested[?index(@) == $.key_to_find]`, and the result is `"real_value"` and the only awareness of `"value_of_key"` is inside that filter selector.

What I'm talking about is this:
```json
{ 
    "a" : 1,
    "value_to_find" : "value_of_value",
    "nested" : {
         "real_key" : "value_of_value",
         "b" : 2
    }
}
```
and I want to reference `"real_key"`. So while I could (I think) do `$.nested[? @ == $.value_to_find]` to get the right node, this results in `"value_of_value"`, not `"real_key"`. The strawman I provided above would be that I'd use `$.nested[? @ == $.value_to_find]#index` to result in `"real_key"`.

**Comment by gregsdennis** [comment]
Yes, I see the distinction now.  Thank you.

I think the SO question asker actually has a use case that's more in line with what's currently in the spec, namely functions _inside expressions_.

**Comment by benkehoe** [comment]
Agreed, a function to return the index/key for a value should satisfy the use case in the StackOverflow question, which I do think is a common one and worth addressing!

**Comment by goessner** [comment]
We already discussed this insufficiency of JSONPath a number of times. From my point of view this would be a useful function.

**Comment by benkehoe** [comment]
> this would be a useful function

"useful function" meaning an `index()` function for filter selectors, or a new selector that allowed referencing object keys (or something else along those lines)? If the latter, should this issue be re-opened?

**Comment by timbray** [comment]
Even better, someone could do a PR.

**Comment by benkehoe** [comment]
Are people open to an "index name selector"? I'd be happy to write a PR for it, but it feels controversial enough that I'd want to get buy-in before getting down to it.

I understand less about the function extensions, I'm probably not the right person to write that PR.

**Comment by cabo** [comment]
We have what would be an index name selector (two actually: `['name']` and `.name`).
Maybe we need an example to understand what you are trying to do.

**Comment by danielaparker** [comment]
> We have what would be an index name selector (two actually: `['name']` and `.name`). 

 `['name']` and `.name` select values, not names.  @benkehoe may have in mind something like the `~` operator in [JSONPath-Plus](https://github.com/JSONPath-Plus/JSONPath). For example, given document

```
{ 
    "a" : 1,
    "value_to_find" : "value_of_value",
    "nested" : {
         "real_key" : "value_of_value",
         "b" : 2
    }
}
```
and query
```
$.nested[*]~
```
JSONPath-Plus returns

```
[
  "real_key",
  "b"
]
```

**Comment by goessner** [comment]
I have two examples in mind:

```
{ "theme": "dark",
  "light": "white",
  "dark": "black"
]
```

Now depending on the value of `"theme"`, the query `$[? index(@) == $.theme]` yields the value `"black"` and/or the normalized path `$['dark']`. For `"theme": "light"` the result would have been `"white"`.

Another example ...

```
{ "even": [2,4,6,8],
 "odd": [ 1,3,5,7,9]
}
```
Here the query `$[? index(@[? @ == 5]) == 'odd'] could verify the container of a value given.

**Comment by gregsdennis** [comment]
> `$[? index(@[? @ == 5]) == 'odd']`

I think something like this would solve both use cases that @benkehoe mentioned in [this comment](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/156#issuecomment-1404221537).

**Comment by goessner** [comment]
> and query
> 
> ```
> $.nested[*]~
> ```
> 
> JSONPath-Plus returns
> 
> ```
> [
>   "real_key",
>   "b"
> ]
> ```
> 
> Jayway JSONPath supports similar functionality with a `keys` function.

And JSONPath yields with the query `$.nested[*]` the normalized pathes

```
["$['nested']['real_key']", 
 "$['nested']['b']"
]
```
from which (with some trivial post processing) we can extract the requested keys easily.

**Comment by benkehoe** [comment]
The `~` operator mentioned by @danielaparker is exactly the sort of thing I'm talking about. I do recognize that it may be possible to extract keys from the results of evaluating path on a document, but that doesn't, for example, allow a user to provide a path as input to a system that accepts paths to tell it to use the value of a particular key in the document. That is, while an implementation may make a key available to the user of that implementation, that key is still not addressable with a JSONPath. A solution which makes keys addressable would *also* solve the use case of making them available in filter expressions.

**Comment by goessner** [comment]
> The `~` operator mentioned by @danielaparker is exactly the sort of thing I'm talking about.

I would like to see a use case / example, where a function for use in filters is not sufficient.

**Comment by glyn** [comment]
We have been pretty conservative in the current spec when it comes to innovation. This approach derives from the [WG charter](https://datatracker.ietf.org/doc/charter-ietf-jsonpath/), which includes the following statement:

> The WG will develop a standards-track JSONPath specification that
is technically sound and complete, based on the common semantics
and other aspects of existing implementations.  Where there are
differences, the working group will analyze those differences and
make choices that rough consensus considers technically best, with
an aim toward minimizing disruption among the different JSONPath
implementations.

So, even if we come up with a sound proposal here, I'd want to defer it beyond v1. Just making this statement now so it doesn't come as a surprise later.

**Comment by benkehoe** [comment]
Understood about v1, I'm happy to wait to bring it up again in the future.

The use case is where I, a developer, am using a system that processes JSON documents, and uses JSONPath as input to control aspects of that processing. An example of this is the [Amazon States Language output/result processing](https://states-language.net/#filters) used by AWS Step Functions.

A concrete example might be a version history in a key-value store, where I can batch-retrieve the version records, where the record keys are the version hash id and they contain the sequential version number, along with the "current version" record that identifies the version by hash, so the document that the system is processing is something like:

```json
{
    "__current__": {
        "current_version_number": 3
    },
    "ae497b": {
        "version_number": 1,
        "commit_message": "initial commit"
    },
    "d43799": {
        "version_number": 2,
        "commit_message": "new functionality"
    },
    "ef6217": {
        "version_number": 3,
        "commit_message": "bugfix"
    }
}
```
and I want the system to retrieve the version hash id. Note that AWS Step Functions is a system where many processing steps like the data retrieval and other API calls can be chained together without custom code, using JSONPath to manipulate the inputs and outputs (and the lack of custom code manipulating the JSON state is desired!). So wiring the step that gathers the above document to a step that sends the version hash id somewhere else can only be done if object keys are addressable with JSONPath.

If I was writing code with a library (either because I'm doing all the processing, or because I've inserted a processing step with custom code in the above system—which I want to avoid), I could use the approach mentioned by @goessner, which is to evaluate `$[? @.version_number == $.__current__.current_version_number]` on the document, and leverage some aspect of the result to get the key, whether that's access to the nodelist or parsing result paths or whatever. However, in this case I do not control the actual processing, but am merely able to provide a JSONPath that will be evaluated for a result.

In this situation, the ability to address the key (with a selector like `#index` or `~`) would enable the specification of a JSONPath to input into the system to correctly process JSON documents in the intended manner.

I think the distinction between a user who both creates JSONPaths and processes them with an implementation, and a user who creates JSONPaths and passes them as input to a system (not controlled by the user) that then processes them with an implementation is relevant here as it was in #155 for singular paths: the former doesn't have a strong need for it because there are easy ways around it, but it's much more difficult in the latter.

**Comment by cabo** [comment]
> A concrete example might be a version history in a key-value store

I note that all the information you need is already in the JSON chunk you show.
What is the output that you actually want to get?

**Comment by gregsdennis** [comment]
☝️ just identifying a use case for this.

**Comment by benkehoe** [comment]
> I note that all the information you need is already in the JSON chunk you show.

@cabo The value I want to reference from the JSON document is the current version hash `ef6217`, but what's important in the use case is that it's not that I want that value myself, but rather that I want a system to retrieve the document, obtain the value `ef6217`, and then carry out some task with that value (e.g., sending it to another system), and the mechanism for telling the system how to obtain the value to use for the task is JSON Path. So it's not a situation where "retrieve the document, process the document in code, and get the key" suffices. I, as the person who constructs a JSON Path in this case, am not in control of the code that is using a JSON Path implementation, nor does my own code ever receive or observe the (pre- or post-processed) document. A system that works like this is AWS Step Functions.

**Comment by gregsdennis** [comment]
@benkehoe that's a good scenario.  We need to consider when a JSON Path is being submitted to an API that we don't control.

**Comment by gregsdennis** [comment]
Here's a [StackOverflow question](https://stackoverflow.com/q/75747193/878701) that highlights a really good use case for this.

The proposed solution would be

```jsonpath
$.config[?index(@)==$.language]
```

Currently, there's no way to support what is requested.

**Comment by benkehoe** [comment]
Additionally, the reverse problem for that example, obtaining the language code based on the localized "hello" value, is a use case for a selector like `#index` or `~`, the proposed solution would be setting `$.language` to something like `$.config[?@.hello == $.localizedHello]#index`, where if `$.localizedHello` was `hallo`, it would result in `$.language` be set to `DE`.

**Comment by gregsdennis** [comment]
@benkehoe your case already has a solution (as was previously mentioned) of finding the associated "hello" text and returning the appropriate segment of the associated path.

> where if `$.localizedHello` was `hallo`, it would result in `$.language` be set to DE.

Your example would result in a nodeset that included a single node: (`hallo`, `$['config']['DE']`).  The `DE` is already present and easily obtainable.

**Comment by benkehoe** [comment]
The use case in the linked SO question is for AWS Step Functions, in which access to the nodeset is not available. You are only able to specify that the value of a (singular) JSONPath gets assigned to a particular place in the output document.

**Comment by gregsdennis** [comment]
Once this specification is finished, it's expected that AWS Step Functions would eventually align with it, in which case the nodeset will likely be available (depending on what they choose to expose).

Regardless, we're not concerned with third-party APIs here.  We're concerned with what is possible with what the spec defines.  If third parties choose to hide some of that functionality, that's their decision, and the best we can do is encourage them not to.

**Comment by benkehoe** [comment]
> we're not concerned with third-party APIs here

This feels like a very different response from [your reply last month](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/156#issuecomment-1421587826), and I'm not sure what's changed:
> that's a good scenario. We need to consider when a JSON Path is being submitted to an API that we don't control.

At heart, it seems very strange to me if JSON Path, a system for addressing content in JSON documents, intends not to make all content in a given document addressable. This is most needed when a JSON Path is the medium of exchange (i.e., a parameter in the API to a JSON-processing system), but I think it applies generally. In the document `{"foo": "bar"}` why should I only be able to point to `"bar"` and not to `"foo"`? I understand that an implementation may enable me, in some implementation-specific way, to refer to `"foo"`, but why leave that to some probably awkward or verbose ad hoc interface, and not provide a clean, concise way for it to be an actual Path (in a future version of the spec)?

**Comment by gregsdennis** [comment]
We can be concerned with when a JSON Path is submitted to an API while at the same time recognizing that we can't control what that API does with it.

> In the document `{"foo": "bar"}` why should I only be able to point to `"bar"` and not to `"foo"`?

We're not preventing that.  The path `$[?@=='bar']` doesn't just return `"bar"`; it also returns the location of `"bar"`, `$['foo']`, which contains `"foo"`.  If the API is more concerned with the property name, it has a way to get that.

We can provide the means for the API to get the information it needs.  What we can't do is make the API use that means.  At this stage, I think that's sufficient.

**Comment by lxhunter** [comment]
OFF TOPIC: I don't know if this is appropriate, but thank you for taking my request so serious! 👍🏻  Really appreciated! Is there a way I can sent a round of coffee to you guys?
